File: core/base/tools/commander/workspace_template/Displays/Core/EVS/Scripts/JS/events_init.js
//Script to demonstrate auto-generation pattern of an OPI that is tailored to the event commands of an APP.

importPackage(Packages.org.csstudio.opibuilder.scriptUtil);
importPackage(org.csstudio.opibuilder.properties);
importPackage(org.csstudio.opibuilder.model);

var app_name  =       display.getMacroValue("APP");

//If you want to add more commands(another column), just add the name of the command to this array. Do note that
//you'll need special logic for adding another XML script to the button that fires that command. Hopefully the pattern below
//makes this obvious. 
var commands = ['Reset'];
if(app_name != null && app_name!="")
{
    display.getWidget("app_name").setPropertyValue("text",app_name);
    for(var i =0;i<8;i++)
	{
	    for(var j =0;j<commands.length;j++)
	       {
		   //The following macro is NOT used here. It's just here for documentation purposes.
		   var event_id_macro_input = DataUtil.createMacrosInput(true);
		   event_id_macro_input.put("EVENT_ID", i);
		   event_id_macro_input.put("APP", "TO");

		    var label_widget =          WidgetUtil.createWidgetModel("org.csstudio.opibuilder.widgets.Label");
		    var button_widget =         WidgetUtil.createWidgetModel("org.csstudio.opibuilder.widgets.ActionButton");
		    var actions =               FileUtil.loadXMLFile("/Displays/Core/EVS/Scripts/XML/ResetEvent.xml");
		    var action_property =       new ActionsProperty(AbstractWidgetModel.PROP_ACTIONS, "Actions", WidgetPropertyCategory.Behavior, false);
		    var xml_input_property =    action_property.readValueFromXML(actions);

		    action_property.setPropertyValue(xml_input_property);
		    button_widget.addProperty(action_property);
		    //The "pv_value" property is used as the event id by "ResetEvent" script. This is becasuse non-container widgets such as ActionButton
		    //do not have macros of their own. However they do have pv_values one may use.
		    button_widget.setPropertyValue("pv_value", i);
		    button_widget.setPropertyValue("text", commands[j]);

		   display.addChild(label_widget);
		   display.addChild(button_widget);
		}
	}

}

File: core/base/tools/commander/workspace_template/Displays/Core/EVS/Scripts/Python/events_init.py
"""
This script auto-generates event ids rows for each event id found in the YAML file. It essentially creates LinkingContainer[1]
widgets and uses EventRecord.opi to generates a container widget for each event in airliner configuration.

[1]:https://docs.yamcs.org/yamcs-studio/
"""
# import java packages

from org.csstudio.opibuilder.scriptUtil import PVUtil, ScriptUtil, FileUtil, WidgetUtil, DataUtil, FileUtil
from org.eclipse.swt.graphics import RGB

from com.windhoverlabs.studio.registry import YAMLRegistry, ConfigRegistry

# import python packages
import logging
import sys

# import our own APIs. Not sure if this is the cleanest way of doing this. If we don't do it this way, we might have add
# an __init__.py to every directory in the project. Don't want to hardcode this path either; open to ideas about this.
sys.path.append(FileUtil.workspacePathToSysPath("Displays"))
from Resources.opi_util import util


def main():
    logging.basicConfig()
    logger = logging.getLogger('events_init')
    if util.validate_opi(display) is False:
        logger.warning("OPI is not valid. Ensure that the APP and PROJECT_NAME macros have some kind of "
                       "value.")
        return -1

    registry_path = display.getMacroValue("REGISTRY_PATH")

    registry = YAMLRegistry()

    app_name = display.getMacroValue("APP")
    project_name = display.getMacroValue("PROJECT_NAME")

    events = util.get_events_from_registry(registry, registry_path)

    for event in events:
        event_label = "{}  ({})  ".format(event, events[event]['id'])

        new_event_record = WidgetUtil.createWidgetModel("org.csstudio.opibuilder.widgets.linkingContainer")
        # TODO: Use the enumeration values from Studio to avoid magical strings
        new_event_record.setPropertyValue("opi_file", display.getMacroValue("TEMPLATE_OPI"))
        new_event_record.setPropertyValue("height", 20)
        new_event_record.setPropertyValue("width", 680)
        new_event_record.setPropertyValue("background_color", RGB(255, 255, 255))
        new_event_record.setPropertyValue("border_color", RGB(240, 240, 240))
        new_event_record.setPropertyValue("border_style", 0)
        new_event_record.setPropertyValue("name", event)

        display.getWidget("EventsTable").addChild(new_event_record)

        # Access children of new_event_record only AFTER they are added to the display container
        new_event_record.getChildByName("EventRecordTemplate").getChildByName("EventId").setPropertyValue("text",
                                                                                                          event_label)

        # Create the variables we need to be able to access data from button scripts.
        new_event_record.getChildByName("EventRecordTemplate").getChildByName("DeleteFilter").setPropertyValue(
            "pv_value", str(events[event]['id']))
        new_event_record.getChildByName("EventRecordTemplate").getChildByName("ResetFilter").setPropertyValue(
            "pv_value", str(events[event]['id']))

        event_macros = DataUtil.createMacrosInput(False)
        event_macros.put("EVENTID", str(events[event]['id']))
        new_event_record.getChildByName("EventRecordTemplate").setPropertyValue("macros", event_macros)

    long_name = util.get_long_name_from_registry(registry, registry_path)
    display.getWidget("app_name").setPropertyValue("text", long_name)


main()
